### PR TITLE
Asset group

### PR DESCRIPTION
**Pull request by iboss-ptk**
This PR introduce concept of `scope` for limiter and upcoming rebalancing incentive

`scope` can be either `denom` or `asset_group` which can practically represent source chain or bridge. This is what changed in registering limiter:

```diff
{
  "register_limiter": {
-  "denom": "token1",
+  "scope": "denom::token1",
    "label": "1h",
    "limiter_params": {
      "change_limiter": {
        "window_config": {
          "window_size": "3600000000000",
          "division_count": "5"
        },
        "boundary_offset": "0.2"
      }
    }
  }
}
```

`scope` can be prefixed with either `denom::` or `asset_group::` to differentiate the scope variant.

In order to specify an `asset_group` as scope, it creating new group is requried
```json
{ 
  "create_asset_group": {
    "label": "group0",
    "denom": ["denom0", "denom1"]
  }
}
```

it can also be removed, and the associated limiters will we removed along with it

```json
{ "remove_asset_group": { "label": "group0" } }
```

`mark_corrupted_assets` &  `unmark_corrupted_assets` also changed to `mark_corrupted_scopes` &  `unmark_corrupted_scopes`

The assets that are within the corrupted scopes will be removed from the pool when the amount reaches 0 as before. The reference to the specific asset that is corrupted and reaches 0 will also be removed from `asset_group`. Once asset group emptied, it is also automatically removed.

### CODE REVIEW

File: contracts/transmuter/src/contract.rs [hunk 1]
```diff
@@ -225,67 +226,155 @@ impl Transmuter<'_> {
...
-    /// As a result, the corrupted assets will not allowed to be increased by any means,
-    /// both in terms of amount and weight.
-    /// The only way to redeem other pool asset, is to also redeem the corrupted asset
-    /// with the same pool-defnined value.
     #[sv::msg(exec)]
-    fn mark_corrupted_assets(
+    fn create_asset_group(
         &self,
         ExecCtx { deps, env: _, info }: ExecCtx,
+        label: String,
```
Comment (crnbarr93) [inline contracts/transmuter/src/contract.rs:252]: nit: Can this label be an empty string?

File: contracts/transmuter/src/transmuter_pool/asset_group.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        ensure!(
+            !self.asset_groups.contains_key(&label),
+            ContractError::AssetGroupAlreadyExists {
+                label: label.clone()
+            }
+        );
+
+        // ensure that all denoms are valid pool assets
+        for denom in &denoms {
+            ensure!(
```
Comment (crnbarr93) [inline contracts/transmuter/src/transmuter_pool/asset_group.rs:104]: Is there a DOS vector here if I add an asset group with the same asset a ton of times? Might be worth enforcing uniqueness of elements

**Review by crnbarr93 — APPROVED** [review]

**Review by crnbarr93 — APPROVED** [review]
Few comments otherwise LGTM!

**Comment by iboss-ptk** [comment]
nice catches, will fix them!

File: contracts/transmuter/src/contract.rs (hunk 1 above)
Comment (iboss-ptk) [reply contracts/transmuter/src/contract.rs:252]: shouldn't, will catht that.

**Pull request merged**